include/cyclic_reduction.hxx

```diff
@@ -290,351 +300,373 @@ public:
         ifc(j, 1) = co(j, 4 * i + 1) - beta * ifc(j, 0);
         ifc(j, 0) = co(j, 4 * i);
-	ifc(j, 2) *= -beta;
+        ifc(j, 2) *= -beta;
         // ic columns  {i-1, i, N-1}
-        
+
         // b_u <- b_i - beta*b_u
         ifc(j, 3) = co(j, 4 * i + 3) - beta * ifc(j, 3);
       }
-      
+
       // Calculate lower interface equation
       // Uses next 4 ifc values for this system so have +4 in indexinf
 
       // v_l <- v_(k+1)
       // b_l <- b_{k+1}
-      for(int i=0;i<4;i++)
+      for (int i = 0; i < 4; i++)
         ifc(j, 4 + i) = co(j, 4 + i);
-        
-      for(int i = 2; i < nloc; i++) {
-	
-	if(abs(ifc(j, 4 + 1)) < 1e-10)
-	  throw BoutException("Zero pivot in CyclicReduce::reduce");
-	
+
+      for (int i = 2; i < nloc; i++) {
+
+        if (abs(ifc(j, 4 + 1)) < 1e-10)
+          throw BoutException("Zero pivot in CyclicReduce::reduce");
+
         // alpha <- v_{i,i-1} / v_l,i-1
         T alpha = co(j, 4 * i) / ifc(j, 4 + 1);
-          
+
         // v_l <- v_i - alpha*v_l
- 	ifc(j, 4 + 0) *= -alpha;
-        ifc(j, 4 + 1) = co(j, 4 * i  +  1) - alpha*ifc(j, 4 + 2);
-        ifc(j, 4 + 2) = co(j, 4 * i  +  2);
+        ifc(j, 4 + 0) *= -alpha;
+        ifc(j, 4 + 1) = co(j, 4 * i + 1) - alpha * ifc(j, 4 + 2);
+        ifc(j, 4 + 2) = co(j, 4 * i + 2);
         // columns of ic are {0, i, i + 1}
-          
+
         // b_l <- b_{k + i} - alpha*b_l
-        ifc(j, 4 + 3) = co(j, 4 * i  +  3) - alpha * ifc(j, 4 + 3);   
+        ifc(j, 4 + 3) = co(j, 4 * i + 3) - alpha * ifc(j, 4 + 3);
       }
-      
+
 #ifdef DIAGNOSE
-      output << "Lower: " << ifc(j, 4 + 0) << ", " << ifc(j, 4 + 1) << ", " << ifc(j, 4 + 2) << " : " << ifc(j, 4 + 3) << endl;
-      output << "Upper: " << ifc(j, 0) << ", " << ifc(j, 1) << ", " << ifc(j, 2) << " : " << ifc(j, 3) << endl;
+      output << "Lower: " << ifc(j, 4 + 0) << ", " << ifc(j, 4 + 1) << ", "
+             << ifc(j, 4 + 2) << " : " << ifc(j, 4 + 3) << endl;
+      output << "Upper: " << ifc(j, 0) << ", " << ifc(j, 1) << ", " << ifc(j, 2) << " : "
+             << ifc(j, 3) << endl;
 #endif
     }
 
     // Lower system couples {0, N-1, N}
     // Upper system couples {-1. 0, N-1}
   }
-  
+
   /// Back-solve from x at ends (x1, xn) to obtain remaining values
   /// Coefficients ordered [ns, nloc*(a,b,c,r)]
-  void back_solve(int ns, int nloc, Matrix<T> &co, Array<T> &x1, Array<T> &xn, Matrix<T> &xa) {
+  void back_solve(int ns, int nloc, Matrix<T> &co, Array<T> &x1, Array<T> &xn,
+                  Matrix<T> &xa) {
     // Tridiagonal system, solve using serial Thomas algorithm
     // xa -- Result for each system
     // co -- Coefficients & rhs for each system
-    Array<T> gam(nloc);
-    for(int i=0;i<ns;i++) { // Loop over systems
+    BOUT_OMP(parallel for)
+    for (int i = 0; i < ns; i++) { // Loop over systems
+      Array<T> gam(nloc);
       T bet = 1.0;
-      xa(i, 0) = x1[i]; // Already know the first 
+      xa(i, 0) = x1[i]; // Already know the first
```

Comment:
Is there any risk of the mentioned race/memory issue happening here such that we need to make this unique outside the parallel region?